snuba/pipeline/query_pipeline.py

```diff
@@ -111,15 +113,51 @@ class QueryPipelineStage(Generic[Tin, Tout]):
         >>>             return QueryPipelineResult(None, e)
     """
 
+    def _process_error(
+        self, pipe_input: QueryPipelineError[Tin]
+    ) -> Union[Tout, Exception]:
+        """default behaviour is to just pass through to the next stage of the pipeline
+        Can be overridden to do something else"""
+        logging.exception(pipe_input.error)
+        return pipe_input.error
+
     @abstractmethod
-    def _execute(self, input: QueryPipelineResult[Tin]) -> QueryPipelineResult[Tout]:
+    def _process_data(self, pipe_input: QueryPipelineData[Tin]) -> Tout:
         raise NotImplementedError
 
-    def execute(self, input: QueryPipelineResult[Tin]) -> QueryPipelineResult[Tout]:
-        if input.error:
-            # Forward the error to next stage of pipeline
-            return QueryPipelineResult(None, input.error)
-        return self._execute(input)
+    def execute(
+        self, pipe_input: QueryPipelineResult[Tin]
+    ) -> QueryPipelineResult[Tout]:
+        if pipe_input.error:
+            res = self._process_error(pipe_input.as_error())
```

Comment:
It's very confusing that `pipe_input.error` is true, and then you process that error `as_error`, but the output is not necessarily an error, and in fact might be a valid result.

Reply:
I can see why you say that but let me give a counter-example. When we report the results of a query to the querylog, we take in either the error or the query result, but in either case, the reporting is successful

Reply:
That's true, what's confusing me is that there are also two classes: `QueryPipelineError` and `QueryPipelineResult`. Those types seem interchangeable? Somehow `pipe_input` is both? This makes sense if `QueryPipelineResult` contains an optional error/result, but it is confusing that the same object `res` can be either an error or a result. There's another layer of confusion that `_process_error`, which would seem to be exclusive to handling errors, can also sometimes output a non-error.